### PR TITLE
fix: App still sends analytics data after analytics is disabled - WPB-10714

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10714" title="WPB-10714" target="_blank">BugWPB-10714</a>  [iOS] App still sends analytics data after analytics is disabled
  </td></table>
  <br />
 

This PR addresses an issue where events were still being sent to Countly even when the user had disabled analytics. Previously, the analytics manager was initialized regardless of whether the user had analytics enabled or disabled. This behavior led to events being tracked even when analytics were supposed to be turned off.

With this fix:

- `EnableAnalyticsUseCase`: The analytics manager is now only initialized when the user explicitly enables analytics.
- `DisableAnalyticsUseCase`: The analytics manager is set to nil when the user disables analytics, ensuring no events are tracked when analytics are turned off.

This approach guarantees that no analytics manager is available when the user opts out, effectively preventing any further event tracking.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

### CODE REVIEW

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift
Comment (netbe) [inline wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: do we need to enable tracking yet if the user logged out? Is it useful for the caller to know if it was successful or not. In that case we could pass a bool in the completion

Comment (agisilaos) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: > do we need to enable tracking yet if the user logged out?

@netbe No, which brings the question do we need the completion there after all. cc: @caldrian

Comment (netbe) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: @agisilaos well I guess you wanted to execute the tracking after the session is loaded

Comment (agisilaos) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: That was the intention yes, because fore that it wouldn't be possible.

Comment (agisilaos) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: @netbe Done https://github.com/wireapp/wire-ios/pull/1888/commits/80a8fed39470c67e77190d5dceea7d573e8a96db

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift
Comment (caldrian) [inline wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: Please run the formatter, it will fix it automatically.

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift
Comment (caldrian) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: I'm not sure we handle login/logout/multiple accounts gracefully.

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: Done!

File: wire-ios-sync-engine/Source/SessionManager/SessionManager.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager.swift]: I already done that on a previous ticket if I remember correctly. Making sure the analytics user profile is updated and we start a new session if there's a different user e.g has multiple accounts.

File: wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift
Comment (KaterinaWire) [inline wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift]: Where do we use `didDisableAnalyticsClosure`? I didn't find in the code

File: wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift
Comment (KaterinaWire) [inline wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift:38]: We are passing 2 huge objects to `DisableAnalyticsUseCase`: `sessionManager` and `userSession`? Is it so necessary? Especially that `sessionManager` already has a reference to the `activeUserSession`.

Comment (KaterinaWire) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift:38]: I can guess that you wanted to pass `didDisableAnalyticsClosure` instead of `sessionManager`, no?

Comment (KaterinaWire) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift:38]: Same question about `EnableAnalyticsUseCase`.  @netbe  and @caldrian  I would like to ask you opinions as well

File: wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift
Comment (caldrian) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift]: It was left-over, deleted now.

File: wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift
Comment (caldrian) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift:38]: > We are passing 2 huge objects to DisableAnalyticsUseCase: sessionManager and userSession?

I don't get the point about "huge objects".
The use cases depend on protocols, one which declares an `analyticsManager` property, where the `AnalyticsManager` reference is set or cleared, and another protocol where the `analyticsSession` can be stored.
It doesn't depend on anything from `ZMUserSession` or `SessionManager`.

If/why the two instances need to be stored separately as well as any implementation detail around `WireAnalytics` I haven't looked into yet.

> Especially that sessionManager already has a reference to the activeUserSession.

Is there always one user session at a time which is available through `SessionManager. activeUserSession`? Then it might make sense to combine the protocols.

Comment (netbe) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift:38]: > I can guess that you wanted to pass didDisableAnalyticsClosure instead of sessionManager, no?
yes that was a way I suggested to not mock userSession in SyncEngine

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (samwyndham) [inline wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: `DisableAnalyticsUseCaseUserSession` & `EnableAnalyticsUseCaseUserSession` are identical right? It is also not clear from their names what they do. I think you should combine them into a single protocol. If they are really needed I think you should combine them into a single protocol and rename it to `AnalyticsSessionProviding`

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (samwyndham) [inline wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: I don't think you need generics here. In the initializer you can just do:

`userSession: DisableAnalyticsUseCaseUserSession`

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (samwyndham) [inline wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: Consider renaming `userSession` here and else where to something like `analyticsSessionProvider` as we don't really care about the `userSession` functionality - only that it provides the analytics session. It could be any object.

File: wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift
Comment (samwyndham) [inline wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift:36]: Same comments in this file as I left in `EnableAnalyticsUseCase`

File: wire-ios-sync-engine/Tests/Source/Use cases/EnableAnalyticsUseCaseTests.swift
Comment (samwyndham) [inline wire-ios-sync-engine/Tests/Source/Use cases/EnableAnalyticsUseCaseTests.swift]: Will you complete this in this PR?

File: wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift
Comment (caldrian) [reply wire-ios-sync-engine/Source/SessionManager/SessionManager+AnalyticsUseCases.swift:38]: The closure is gone.

File: wire-ios-sync-engine/Tests/Source/Use cases/EnableAnalyticsUseCaseTests.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Tests/Source/Use cases/EnableAnalyticsUseCaseTests.swift]: Done: https://github.com/wireapp/wire-ios/pull/1888/commits/e0ee1127c879ce3173717e82559483f4b54a33fe

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (samwyndham) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: I guess you are using Generics to have static vs dynamic dispatch like @caldrian in some of his PRs. Is that correct?

**Review by samwyndham — COMMENTED** [review]
@agisilaos There are some comments I would like to understand before I continue review. Once those are answered I will complete the review. Thank you!

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: Removed the generics from the use cases: https://github.com/wireapp/wire-ios/commit/c82f5d3b1e3b7847902aa182978145fcd4e241d5

File: wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift:36]: Removed the generics from the use cases: https://github.com/wireapp/wire-ios/commit/c82f5d3b1e3b7847902aa182978145fcd4e241d5

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: Done: https://github.com/wireapp/wire-ios/commit/97b486e3536ba60882aa7ac9a80cea7c6f76e7de

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: I thought about it and I would like to keep those two separate even though they're containing the same property so we have better separation of concerns if we decide to add anything else inside those protocols.

Comment (samwyndham) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: 👍 Yeah I guess there are always tradeoffs - I would probably take the other approach but it is all subjective! 

Maybe still consider renaming as they don't explicitly need a user session. For example, `DisableAnalyticsUseCaseAnalyticsSession`

**Review by samwyndham — APPROVED** [review]
Nice work. Approving with the knowledge that you are still working on the UI toggle issue.

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (caldrian) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift]: optional
```suggestion
protocol DisableAnalyticsUseCaseAnalyticsSessionProviding: AnyObject {
```

File: wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift
Comment (caldrian) [inline wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift]: optional
```suggestion
protocol EnableAnalyticsUseCaseAnalyticsSessionProviding: AnyObject {
```

File: wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift
Comment (caldrian) [inline wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift:66]: Since the `init` doesn't have to be `public`, it can be omitted:
Just declare the properties like this:
```
/// The configuration for the analytics session.
private(set) var analyticsSessionConfiguration: AnalyticsSessionConfiguration

/// A closure that provides an `AnalyticsManagerProtocol` implementation.
private(set) var analyticsManagerBuilder: (_ appKey: String, _ host: URL) -> any AnalyticsManagerProtocol

/// The session manager that conforms to `AnalyticsManagerProviding` for managing analytics sessions.
private(set) var sessionManager: AnalyticsManagerProviding

/// The user profile for which to enable analytics sharing.
private(set) var analyticsUserProfile: AnalyticsUserProfile

/// An instance conforming to `EnableAnalyticsUseCaseUserSession` that manages the user's analytics session state.
private(set) var analyticsSessionProvider: EnableAnalyticsUseCaseUserSession
```

**Review by caldrian — APPROVED** [review]

File: wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift]: Done!

File: wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift
Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/EnableAnalyticsUseCase.swift:66]: Done!

File: wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift
Comment (johnxnguyen) [inline wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift:47]: Why wasn't this enough before? It seems like a bug that even if this is called, that the events are still being sent, regardless of the `AnalyticsManager` existing.

Comment (netbe) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift:47]: @agisilaos ?

Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift:47]: @johnxnguyen Not really sure to be honest. As soon as we took the analytics manager out of the equation we stopped receiving events when disabled. After we merge everything to the feature branch. I want to take another look at the whole refactor and set of changes as a good practice to see what we can improve. 

For now is it ok to go ahead and merge it?

Comment (johnxnguyen) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift:47]: Your solution introduces some complexity that would be nice to avoid. Since this is somewhat urgent, I'm happy to approve it but I think the time should be taken to get to the bottom of the issue.

Comment (johnxnguyen) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift:47]: But then... if we know that `disableAnayltics()` doesn't work, why do we still call it?

Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift:47]: Sounds good! 👍

Comment (agisilaos) [reply wire-ios-sync-engine/Source/Use cases/DisableAnalyticsUseCase.swift:47]: We're going to revisit with @johnxnguyen after everything is merged.

**Pull request merged**
